File: electron/sync/syncEngine.ts
import type { PrismaClient } from '../../generated/prisma/client'
import type { SupabaseClient } from '@supabase/supabase-js'

type SyncStatus = 'idle' | 'syncing' | 'error' | 'offline'

// camelCase → snake_case field mapping for Supabase
const TABLE_MAP: Record<string, string> = {
  Company:     'companies',
  Aircraft:    'aircraft',
  Flight:      'flights',
  Transaction: 'transactions',
  Route:       'routes',
  Dispatch:    'dispatches',
  CrewMember:  'crew_members',
  Loan:        'loans',
  GameEvent:   'game_events',
  Reputation:  'reputations',
}

function camelToSnake(str: string): string {
  return str.replace(/[A-Z]/g, (c) => `_${c.toLowerCase()}`)
}

function toSnakeKeys(obj: Record<string, unknown>): Record<string, unknown> {
  const result: Record<string, unknown> = {}
  for (const [key, value] of Object.entries(obj)) {
    result[camelToSnake(key)] = value
  }
  return result
}

export function createSyncEngine(prisma: PrismaClient, supabase: SupabaseClient, userId: string) {
  let status: SyncStatus = 'idle'
  let timer: ReturnType<typeof setInterval> | null = null
  let onStatusChange: ((s: SyncStatus) => void) | null = null

  function setStatus(s: SyncStatus) {
    status = s
    onStatusChange?.(s)
  }

  async function pushNow(): Promise<void> {
    if (status === 'syncing') return

    // Check connectivity
    try {
      const { error } = await supabase.from('profiles').select('id').limit(1)
      if (error) { setStatus('offline'); return }
    } catch {
      setStatus('offline')
      return
    }

    setStatus('syncing')

    try {
      const pending = await prisma.syncLog.findMany({
        where: { syncedAt: null },
        orderBy: { createdAt: 'asc' },
        take: 200,
      })

      if (pending.length === 0) {
        setStatus('idle')
        return
      }

      // Group by table for batch operations
      const grouped = new Map<string, typeof pending>()
      for (const entry of pending) {
        const list = grouped.get(entry.tableName) ?? []
        list.push(entry)
        grouped.set(entry.tableName, list)
      }

      for (const [tableName, entries] of grouped) {
        const supabaseTable = TABLE_MAP[tableName]
        if (!supabaseTable) continue

        // Handle deletes
        const deletes = entries.filter((e) => e.action === 'delete')
        if (deletes.length > 0) {
          const ids = deletes.map((e) => e.recordId)
          await supabase.from(supabaseTable).delete().in('id', ids)
        }

        // Handle creates and updates (upsert)
        const upserts = entries.filter((e) => e.action !== 'delete' && e.payload)
        if (upserts.length > 0) {
          const BATCH_SIZE = 50
          for (let i = 0; i < upserts.length; i += BATCH_SIZE) {
            const batch = upserts.slice(i, i + BATCH_SIZE)
            const rows = batch.map((e) => {
              const data = JSON.parse(e.payload!)
              const snakeData = toSnakeKeys(data)
              snakeData.user_id = userId
              return snakeData
            })
            await supabase.from(supabaseTable).upsert(rows, { onConflict: 'id' })
          }
        }

        // Mark as synced
        const ids = entries.map((e) => e.id)
        await prisma.syncLog.updateMany({
          where: { id: { in: ids } },
          data: { syncedAt: new Date() },
        })
      }

      // Update syncedAt on the local records
      const now = new Date()
      for (const entry of pending) {
        if (entry.action === 'delete') continue
        try {
          const model = (prisma as any)[entry.tableName.charAt(0).toLowerCase() + entry.tableName.slice(1)]
          if (model?.update) {
            await model.update({
              where: { id: entry.recordId },
              data: { syncedAt: now },
            })
          }
        } catch {
          // Record may have been deleted locally — ignore
        }
      }

      setStatus('idle')
    } catch (err) {
      console.error('[Sync] Push failed:', err)
      setStatus('error')
    }
  }

  async function pullOnStartup(): Promise<void> {
    setStatus('syncing')

    try {
      for (const [prismaModel, supabaseTable] of Object.entries(TABLE_MAP)) {
        // Get the latest syncedAt from local
        const modelKey = prismaModel.charAt(0).toLowerCase() + prismaModel.slice(1)
        const model = (prisma as any)[modelKey]
        if (!model?.findFirst) continue

        const latest = await model.findFirst({
          where: { syncedAt: { not: null } },
          orderBy: { syncedAt: 'desc' },
          select: { syncedAt: true },
        })

        const since = latest?.syncedAt ?? new Date(0)

        // Pull records updated after our last sync
        const { data, error } = await supabase
          .from(supabaseTable)
          .select('*')
          .gt('updated_at', since.toISOString())
          .order('updated_at', { ascending: true })
          .limit(500)

        if (error || !data?.length) continue

        // Upsert into local DB
        for (const row of data) {
          const camelData = snakeToCamelKeys(row)
          camelData.syncedAt = new Date()
          delete camelData.userId // userId is stored as user_id in cloud
          try {
            await model.upsert({
              where: { id: camelData.id },
              create: camelData,
              update: camelData,
            })
          } catch {
            // Schema mismatch or constraint violation — skip
          }
        }
      }

      setStatus('idle')
    } catch (err) {
      console.error('[Sync] Pull failed:', err)
      setStatus('error')
    }
  }

  function start() {
    if (timer) return
    timer = setInterval(() => pushNow().catch(console.error), 60_000)
  }

  function stop() {
    if (timer) {
      clearInterval(timer)
      timer = null
    }
  }

  function getStatus(): SyncStatus {
    return status
  }

  function onStatus(cb: (s: SyncStatus) => void) {
    onStatusChange = cb
  }

  return { start, stop, pushNow, pullOnStartup, getStatus, onStatus }
}

function snakeToCamelKeys(obj: Record<string, unknown>): Record<string, any> {
  const result: Record<string, any> = {}
  for (const [key, value] of Object.entries(obj)) {
    const camelKey = key.replace(/_([a-z])/g, (_, c) => c.toUpperCase())
    result[camelKey] = value
  }
  return result
}

File: legacy/Thrustline.WPF/Thrustline/Services/Sync/SyncEngine.cs
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Text.RegularExpressions;
using CommunityToolkit.Mvvm.Messaging;
using Microsoft.EntityFrameworkCore;
using Thrustline.Data;
using Thrustline.Messages;

namespace Thrustline.Services.Sync;

public class SyncEngine : IDisposable
{
    private readonly IDbContextFactory<ThrustlineDbContext> _dbFactory;
    private readonly HttpClient _http;
    private readonly string _supabaseUrl;
    private readonly string _anonKey;
    private string? _accessToken;
    private string? _userId;
    private Timer? _timer;
    private string _status = "idle";

    private static readonly Dictionary<string, string> TableMap = new()
    {
        ["Company"] = "companies", ["Aircraft"] = "aircraft", ["Flight"] = "flights",
        ["Transaction"] = "transactions", ["Route"] = "routes", ["Dispatch"] = "dispatches",
        ["CrewMember"] = "crew_members", ["Loan"] = "loans", ["GameEvent"] = "game_events",
        ["Reputation"] = "reputations",
    };

    public string Status => _status;

    public SyncEngine(IDbContextFactory<ThrustlineDbContext> dbFactory, string supabaseUrl, string anonKey)
    {
        _dbFactory = dbFactory;
        _supabaseUrl = supabaseUrl.TrimEnd('/');
        _anonKey = anonKey;
        _http = new HttpClient();
    }

    public void Configure(string accessToken, string userId)
    {
        _accessToken = accessToken;
        _userId = userId;
    }

    public void Start()
    {
        _timer ??= new Timer(async _ => await PushNowAsync(), null, TimeSpan.FromSeconds(60), TimeSpan.FromSeconds(60));
    }

    public void Stop()
    {
        _timer?.Dispose();
        _timer = null;
    }

    private void SetStatus(string s)
    {
        _status = s;
        WeakReferenceMessenger.Default.Send(new SyncStatusMessage(s));
    }

    public async Task PushNowAsync()
    {
        if (_status == "syncing" || _accessToken == null || _userId == null) return;

        SetStatus("syncing");
        try
        {
            await using var db = await _dbFactory.CreateDbContextAsync();

            var pending = await db.SyncLogs
                .Where(s => s.SyncedAt == null)
                .OrderBy(s => s.CreatedAt)
                .Take(200)
                .ToListAsync();

            if (pending.Count == 0) { SetStatus("idle"); return; }

            var grouped = pending.GroupBy(e => e.TableName);

            foreach (var group in grouped)
            {
                if (!TableMap.TryGetValue(group.Key, out var supaTable)) continue;

                // Deletes
                var deletes = group.Where(e => e.Action == "delete").ToList();
                if (deletes.Count > 0)
                {
                    var ids = string.Join(",", deletes.Select(e => $"\"{e.RecordId}\""));
                    var req = CreateRequest(HttpMethod.Delete, $"/rest/v1/{supaTable}?id=in.({ids})");
                    await _http.SendAsync(req);
                }

                // Upserts
                var upserts = group.Where(e => e.Action != "delete" && e.Payload != null).ToList();
                for (var i = 0; i < upserts.Count; i += 50)
                {
                    var batch = upserts.Skip(i).Take(50).Select(e =>
                    {
                        var data = JsonSerializer.Deserialize<Dictionary<string, JsonElement>>(e.Payload!);
                        var snake = new Dictionary<string, object?>();
                        if (data != null)
                            foreach (var (k, v) in data)
                                snake[CamelToSnake(k)] = v;
                        snake["user_id"] = _userId;
                        return snake;
                    }).ToList();

                    var json = JsonSerializer.Serialize(batch);
                    var req = CreateRequest(HttpMethod.Post, $"/rest/v1/{supaTable}");
                    req.Headers.Add("Prefer", "resolution=merge-duplicates");
                    req.Content = new StringContent(json, Encoding.UTF8, "application/json");
                    await _http.SendAsync(req);
                }

                // Mark synced
                var ids2 = group.Select(e => e.Id).ToList();
                await db.SyncLogs.Where(s => ids2.Contains(s.Id)).ExecuteUpdateAsync(s => s.SetProperty(x => x.SyncedAt, DateTime.UtcNow));
            }

            SetStatus("idle");
        }
        catch
        {
            SetStatus("error");
        }
    }

    private HttpRequestMessage CreateRequest(HttpMethod method, string path)
    {
        var req = new HttpRequestMessage(method, $"{_supabaseUrl}{path}");
        req.Headers.Add("apikey", _anonKey);
        if (_accessToken != null)
            req.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _accessToken);
        return req;
    }

    private static string CamelToSnake(string str) =>
        Regex.Replace(str, "([A-Z])", "_$1").ToLowerInvariant().TrimStart('_');

    public void Dispose()
    {
        Stop();
        _http.Dispose();
        GC.SuppressFinalize(this);
    }
}
